### PR TITLE
Replace ASSERT with VERIFY

### LINKED ISSUE #151: Assert to Verify
We should replace the standard asserts in the code with a runtime verify functionality. We use them as actual error handlers, not only as debug asserts.

This allows us to keep them by default enabled in release builds, making errors way more usable to users.

As an example, [libPMacc](https://github.com/ComputationalRadiationPhysics/picongpu/blob/9b996254c4f6a667c91d119aa5ee9715b21d0912/include/pmacc/verify.hpp) has a verify function.

### PR DESCRIPTION
**Pull request by anokfireball**
Closes #151.

### CODE REVIEW

**Comment by anokfireball** [comment]
This error was present [before](https://ci.appveyor.com/project/ax3l/openpmd-api/build/1.0.169/job/hccu4numdacfqnrd?fullLog=true#L2785), but successfully hidden as the ```ASSERT``` was disabled in ```Release``` builds:
The path is hard-coded to ```./working/directory/2D_simData.h5```, which natrually can't be created on Windows.

File: src/Series.cpp [hunk 1]
```diff
@@ -64,18 +64,33 @@ std::string cleanFilename(std::string const& filename, Format f);
...
+    }
+#endif
     std::string newPath;
     std::string newName;
-    auto const pos = filepath.find_last_of('/');
+    auto const pos = filepath.find_last_of(auxiliary::directory_separator);
     if( std::string::npos == pos )
     {
-        newPath = "./";
+        newPath = "." + auxiliary::directory_separator;
```
Comment (ax3l) [inline src/Series.cpp]: Clang 5 warns `-Wstring-plus-int`:
```
'const char' to a string does not append to the string
(use array indexing to silence this warning)
```

me: explicitly cast or actually `.append()` as below to silence this warning ;-)

Comment (anokfireball) [reply src/Series.cpp]: I made that exact [change](https://github.com/C0nsultant/libopenPMD/blob/0b97c08c0094bc52f461d15d3d905e43b7ac44a6/src/Series.cpp#L173-L174) in the non-MPI constructor, but forgot to copy-paste...

The constructor looks like a good candidate for refactoring, anyway.

**Comment by anokfireball** [comment]
I'll add a changelog later today. Then this should be ready to merge.

**Comment by ax3l** [comment]
alright, just remove the `[WIP]` to indicate when you are finished working on it

File: CMakeLists.txt [hunk 2]
```diff
@@ -45,26 +45,24 @@ function(openpmd_option name description default)
...
+openpmd_option(ADIOS2         "Enable ADIOS2 support"                     OFF)
+#openpmd_option(JSON           "Enable JSON support"                      AUTO)
+openpmd_option(INVASIVE_TESTS "Enable unit tests that modify source code" AUTO)
+openpmd_option(PYTHON         "Enable Python bindings"                    AUTO)
 
 option(openPMD_USE_INTERNAL_VARIANT "Use internally shipped MPark.Variant" ON)
-option(openPMD_USE_INTERNAL_CATCH   "Use internally shipped Catch2" ON)
+option(openPMD_USE_INTERNAL_CATCH   "Use internally shipped Catch2"        ON)
+
+option(openPMD_ENABLE_VERIFY "Enable internal VERFIY macro" ON)
```
Comment (ax3l) [inline CMakeLists.txt]: - typo: "verify" macro
- can you use `openPMD_USE_VERIFY` so we have it consistently named?
- can we write "Keep VERIFY (assert) macros enabled in release mode"?

File: README.md [hunk 3]
```diff
@@ -182,10 +182,12 @@ CMake controls options with prefixed `-D`, e.g. `-DopenPMD_USE_MPI=OFF`:
 | `openPMD_USE_ADIOS2`         | AUTO/ON/**OFF**  | Enable support for ADIOS2 <sup>1</sup>                 |
 | `openPMD_USE_PYTHON`         | **AUTO**/ON/OFF  | Enable Python bindings                                 |
 | `openPMD_USE_INVASIVE_TESTS` | **AUTO**/ON/OFF  | Enable unit tests that modify source code <sup>2</sup> |
+| `openPMD_ENABLE_VERIFY`      | **ON**/OFF       | Enable internally used VERIFY macro <sup>3</sup>       |
```
Comment (ax3l) [inline README.md]: if changed above: name & description

File: docs/source/dev/buildoptions.rst [hunk 4]
```diff
@@ -20,13 +20,16 @@ CMake Option                   Values          Description
 ``openPMD_USE_ADIOS2``         AUTO/ON/**OFF** Enable support for ADIOS2 :sup:`1`
 ``openPMD_USE_PYTHON``         **AUTO**/ON/OFF Enable Python bindings
 ``openPMD_USE_INVASIVE_TESTS`` **AUTO**/ON/OFF Enable unit tests that modify source code :sup:`2`
+``openPMD_ENABLE_VERIFY``      **ON**/OFF      Enable internally used VERIFY macro :sup:`3`
```
Comment (ax3l) [inline docs/source/dev/buildoptions.rst]: if changed above: name & description

File: include/openPMD/Series.hpp [hunk 5]
```diff
@@ -50,11 +50,11 @@ class Series : public Attributable
 
 public:
 #if openPMD_HAVE_MPI
-    Series(std::string const& filepath,
+    Series(std::string filepath,
```
Comment (ax3l) [inline include/openPMD/Series.hpp:53]: no const ref anymore? :'(

**Review by ax3l — APPROVED** [review]

**Comment by ax3l** [comment]
@C0nsultant in `ParallelADIOS1IOHandler.cpp` there are still a few `ASSERT` calls which crash on compile. Did you forget to include the proper header for those asserts or did you forget to change them to `VERIFY`?

**Comment by anokfireball** [comment]
@ax3l I rebased #258 before merging. Apparently not all conflicts were resolved as intended... 
Mea culpa.

**Comment by ax3l** [comment]
@C0nsultant no prob, I just wonder: shall I replace them with `VERIFY` or add the missing include?

**Comment by anokfireball** [comment]
@ax3l Yeah, they should definitely be ```VERIFY``` now.
Sidenote - all the more reason to have ADIOS in CI. Preferably in both serial and parallel. #190

**Comment by ax3l** [comment]
@C0nsultant cool, will do

**Pull request merged**